### DISCUSSION TITLE
Any tips for Raspberry Ri Zero?

### DISCUSSION THREAD

**Discussion by eigenphase** [merbanan/rtl_433#2469] [Q&A]
I got a Raspberry Pi Zero WH and the PoE hat which I want to use as "RF gateway" (both zigbee2mqtt and rtl_433 but for now I start with rtl_433 exclusively). I have connected a cheap "RTL2838 DVB-T".

I installed Raspberry Pi OS, removed all bulk/crap I could think of and compiled rtl_433 from scratch.

Now the issue is, this works mostly. Most of the time. If I run it for long, it seems not everything is always received. For example, I use it for DSC security sensors (window/door) and sometimes when opening/closing, no update is sent. The sensors also provide a "heart beat" message every few hours but sometimes none is received.

I also have Acurite temperature sensors. Since they transmit updates so frequently I don't see the issue there (a few dropped frames are not even visible to me).

At least while I am looking, rtl_433 uses between 4 and 10% CPU. Total load 10-20% (with the remaining part mostly htop, screen, ssh; I removed everything but udiskd, polkitd, rngd, thd, rsyslogd, dhcpd, cron).

Currently I am using this as my command line:

`rtl_433 -F mqtt://10.227.79.10,user=rtl433,pass=xxx,retain=0 -F kv -M newmodel -M time:unix:usec:utc -M level -M noise -M stats  -C si -R 23 -R 11   -b 524288  # -R 209`

Has anyone used rtl_433 on RPiZero and can share some advice?

What could I tweak to make things more reliable?

Is the "-b" option something that I should use or not for increased reliability?

Can I figure out somehow if the issue is poor RF performance versus some low-level driver issues (e.g., dropped data) versus some rtl_433 issue? For example, can it be that Pi's CPU is too poor and sometimes data is can be missed?

Noise is always between -10dB and -11dB:

`[Auto Level] Current noise level -11.4 dB, estimated noise -11.4 dB`

and SNR seems to be consistently ~11dB (no matter which sensor, if near or far). I assume the "Auto Level" assumes signal as 0dB and sets gain accordingly....

I already tried connecting with a USB extension cord. Initially SNR went up to +20dB (if I recall correctly) but subsequently dropped back to ~10dB where it was with adapter directly connected. Overall, I feel connecting via USB cable has improved things slightly but it's still not reliable and I don't get all open/close messages from my DSC Security Sensors...

**Comment by eigenphase** [comment]
To add: It seems rtl_433 literally gets stuck. All of the sudden no new messages. rtl_433 have 0% CPU in htop and this is the last thing on the output (don't know for how long nothing happened):

```
model     : Acurite-609TXC                         id        : 20
Battery   : 1            Temperature: 9.6 C        Humidity  : 60 %          status    : 2             Integrity : CHECKSUM
Modulation: ASK          Freq      : 434.0 MHz
RSSI      : -0.1 dB      SNR       : 11.7 dB       Noise     : -11.8 dB
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 1680574288.964989
model     : Acurite-609TXC                         id        : 20
Battery   : 1            Temperature: 9.6 C        Humidity  : 60 %          status    : 2             Integrity : CHECKSUM
Modulation: ASK          Freq      : 434.0 MHz
RSSI      : -0.1 dB      SNR       : 11.4 dB       Noise     : -11.5 dB
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 1680574289.647073                      enabled   : 2             since     : 2023-04-04T02:01:34                     frames    :
count     : 333          fsk       : 0             events    : 331           stats     :
device    : 23           name      : DSC Security Contact                    events    : 17            ok        : 16            messages  : 16
abort_early: 1,
device    : 11           name      : Acurite 609TXC Temperature and Humidity Sensor                    events    : 335           ok        : 316
messages  : 316          abort_length: 19
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 1680582958.940996                      enabled   : 2             since     : 2023-04-04T04:35:58                     frames    :
count     : 0            fsk       : 0             events    : 0             stats     :
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 1680582959.038317                      enabled   : 2             since     : 2023-04-04T04:35:59                     frames    :
count     : 0            fsk       : 0             events    : 0             stats     :

```

**Reply by zuckschwerdt** [reply]
This might be #2426
The `-M newmodel` is ancient and deprecated, are you running a recent version?
11 dB SNR is not enough and sounds like a jammed band. Observe the band and check it's free. (maybe try `-F rtl_tcp`

**Reply by eigenphase** [reply]
Indeed. #2426 could be related. Having said that, rtl_433 being stuck didn’t occur for a day now; yet messages are still missed a lot.

I am running the newest version from GitHub. Can I just remove -M newmodel without implications?

also I realized I am getting PLL not locked. Could this be an issue?

```

rtl_433 -F mqtt://10.227.79.10,user=rtl433,pass=xxx,retain=0 -F kv -M newmodel -M time:unix:usec:utc -M level -M noise -M stats -C si -R 23 -R 11 -f 433.92M -s 250000 -b 512k # -b 524288  # -R 209
rtl_433 version nightly-21-gfe3df68d branch master at 202303301119 inputs file rtl_tcp RTL-SDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/root/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
MQTT: Publishing MQTT data to 10.227.79.10 port 1883
MQTT: Publishing device info to MQTT topic "rtl_433/rfpi/devices[/type][/model][/subtype][/channel][/id]".
MQTT: Publishing events info to MQTT topic "rtl_433/rfpi/events".
MQTT: Publishing states info to MQTT topic "rtl_433/rfpi/states".
newmodel option (-M) is deprecated.
[Protocols] Registered 2 out of 243 device decoding protocols [ 23 11 ]
Found Rafael Micro R820T tuner
[SDR] Using device 0: Realtek, RTL2838UHIDIR, SN: 00000001, "Generic RTL2832U OEM"
Exact sample rate is: 250000.000414 Hz
[R82XX] PLL not locked!
Allocating 15 zero-copy buffers
```

> 11 dB SNR is not enough and sounds like a jammed band.

what is a typical SNR I should be seeing for (a) a sensor a few ft away no walls and such (b) sensor opposite of the house ? What is deemed “good”, what is deemed “sufficient” and what is deemed “poor”?

also surprisingly when I set gain=auto it always goes to this noise level/SNR, nearly independent where the sensor is. Is this expected?

i tried a few manual gains: gain=45 gave me noise=-21dB. Gain=40 noise=-26dB. Surprisingly RSSI is always around 0dB. But generally I don’t feel performance has improved. Is automatic gain recommend or I should playing with manual gain give me better results?

Similarly should I expect things to get better by playing with -b (or -f, -s) or is this futile?


> Observe the band and check it's free. (maybe try -F rtl_tcp

Good idea. With this you mean plug the adapter into a laptop, pipe data over tcp for rtl_433 to receive?

**Reply by zuckschwerdt** [reply]
> I am running the newest version from GitHub. Can I just remove -M newmodel without implications?

Yes, the option hasn't been doing anything for ~2 years. You should see `newmodel option (-M) is deprecated.`

> also I realized I am getting PLL not locked. Could this be an issue?

No, that's normal for librtlsdr.

> rtl_433 ... -b 512k

If doubling the block does anything at all then it's not good.

> what is a typical SNR I should be seeing for (a) a sensor a few ft away no walls and such (b) sensor opposite of the house ? What is deemed “good”, what is deemed “sufficient” and what is deemed “poor”?

On a clear band expect at worst -20 dB noise floor (20 dB SNR to a 0 dB signal). Usually way more, -30 excellent, cap would be around -36 dB.

We have default limit of minimum signal level (RSSI) of -12 dB and minimum signal noise ratio of 9 dB.
Try `-Y autolevel` on a clear band to lower the minimum RSSI.

> also surprisingly when I set gain=auto it always goes to this noise level/SNR, nearly independent where the sensor is. Is this expected?

No, SNR should always be a function of the signal strength: far away less SNR as the signal is closer to noise.
Auto-gain does not work well for some Fitipower RTL-SDR. I only have R820T(2)'s which work great.

> i tried a few manual gains: gain=45 gave me noise=-21dB. Gain=40 noise=-26dB. Surprisingly RSSI is always around 0dB. But generally I don’t feel performance has improved. Is automatic gain recommend or I should playing with manual gain give me better results?

That's sounds way better. You want signals well below 0 dB which is clipping and impedes decoding. -1 to -9 dB are great for performance.

> Similarly should I expect things to get better by playing with -b (or -f, -s) or is this futile?

Only `-f`, otherwise the default are well tuned.

> > Observe the band and check it's free. (maybe try -F rtl_tcp
> 
> Good idea. With this you mean plug the adapter into a laptop, pipe data over tcp for rtl_433 to receive?

Even easier. Run rtl_433 where it is and add `-F rtl_tcp` then you can use any SDR GUI to connect to it using a rtl_tcp driver.

**Reply by eigenphase** [reply]
> Even easier. Run rtl_433 where it is and add -F rtl_tcp then you can use any SDR GUI to connect to it using a rtl_tcp driver.

Thank you! I'll debug a bit more once I'm home.

> That's sounds way better. You want signals well below 0 dB which is clipping and impedes decoding. -1 to -9 dB are great for performance.

A few hours back I have set gain=35 which gave me ~ -31dB noise. For 3 hrs, my three Acurite-609TXC worked nicely. All of the sudden I saw that I wasn't receiving a message any more from the farthest sensor (id=20). I restarted rte_433, I briefly received some messages (forgot RSSI/Noise level) and then silence from that sensor again. I have now set gain=38 and got:

1. id=20 (farthest sensor; outside, maybe 20m away): RSSI=-12dB, Noise=-27dB
2. id=200 (maybe 5m away): RSSI=-0.8dB, Noise=-30dB
3. id=218 (maybe 1-2 feet away with wooden floor in between): RSSI=-0.1dB, Noise=-29.1dB

RSSI=-12dB I would really call good, no? And with gain=35 (instead of gain=38) this should be -15dB ... which is still good, right? 

This is SO weird! Can a RSSI of -15dB and -31dB noise really be too bad such that all of the sudden nothing received any more? Maybe temperature drift? Clock drift?

**EDIT**: I missed this -- apparently -12dB is indeed considered not good?
> We have default limit of minimum signal level (RSSI) of -12 dB and minimum signal noise ratio of 9 dB. Try -Y autolevel on a clear band to lower the minimum RSSI.

I now set `-g 38 -Y autolevel` and see if this makes any difference. 

Also: For my understanding, what are the actual levels referred to? Is it "dBFS" (i.e., 0dB=0dBFS is ADC full scale) ? In:

```
Auto Level] Current noise level -26.0 dB, estimated noise -17.7 dB
```

is this also "dBFS" ?
What is the difference between noise level, estimated noise and the "Noise" value given by individual sensors?

Before I am getting too frustrated (and wasting resources), can it be that the dongle is just super bad quality? (I got it off eBay for a few bucks 5 years ago...). Is there a "rock solid" dongle that you would recommend so I can get dongle issues out of the way?

[This one](https://www.amazon.com/dp/B01HSSLHZI/) would be intriguing, due to its compact size (but if I would still need to use a USB extension there's no point).

Or is the [official one](https://www.amazon.com/dp/B0BMKZCKTF/) worth the money ?

**Reply by zuckschwerdt** [reply]
> **EDIT**: I missed this -- apparently -12dB is indeed considered not good?

It can be good, but you need `autolevel` and a clear band, i.e. more than 9 dB SNR is always needed and the minimum level needs to be below what you want to receive. (I regularly have a minimum autolevel of -23 dB.)

> Also: For my understanding, what are the actual levels referred to? Is it "dBFS" (i.e., 0dB=0dBFS is ADC full scale) ?

Yes, ADC FS. If you set a fixed level you might be able to construct an absolute reading but actual gains will differ for each hardware. Most people use AGC and then you won't have any absolute reference.

> Auto Level] Current noise level -26.0 dB, estimated noise -17.7 dB

That looks like shortly after startup. The estimation is a slow average and should drift near the current level. It's too far away here.

> is this also "dBFS" ? What is the difference between noise level, estimated noise and the "Noise" value given by individual sensors?

They all capture a different time and the estimate is a slow/longer average.

> Before I am getting too frustrated (and wasting resources), can it be that the dongle is just super bad quality?

Hard to say, I have plenty of the very cheap ones (black or blue plastic sticks) and they work well. You likely always want the R820T (which you have).

> Or is the [official one](https://www.amazon.com/dp/B0BMKZCKTF/) worth the money ?

You get temperature stabilization and a metal case. Nooelec also has some compact ones (NESDR Nano 3). Look for R820T (R860 is a newer name), the E4000 ones might be more complicated.

**Reply by eigenphase** [reply]
> Even easier. Run rtl_433 where it is and add -F rtl_tcp then you can use any SDR GUI to connect to it using a rtl_tcp driver.

Finally got this to try. I also placed the RPIZero+SDR now (easy since it runs over POE) on the other side of the house. All of the sudden sensors that used to be problematic work well (so far) and all sensors that used to work well don't work reliably (or not at all). This suggests noise/signal issue.

Does the noise floor below look ok? It's with `g=36 -Y autolevel -F rtl_tcp:10.227.4.11:1235`.
You can also see one strong transmission that occured (I assume the periodic pattern is harmonics -- for most transmittions it's just a sinble blurb).

![image](https://user-images.githubusercontent.com/7464149/230429326-a58c8ef2-1d54-4ad2-ba38-a5a931ed5434.png)

How do I get noise power from this plot? noise floor seems around -73dBFS. With ~250kHz bandwidth, Pnoise = -73dBFS + 10*log10(250e3) = -19dBFS. However, this is one example with this setting (note SNR and noise still widly vary):

```
time      : 1680796117.838355
model     : Acurite-609TXC                         id        : 20
Battery   : 1            Temperature: 9.7 C        Humidity  : 76 %          status    : 2             Integrity : CHECKSUM
Modulation: ASK          Freq      : 434.0 MHz
RSSI      : -2.6 dB      SNR       : 34.8 dB       Noise     : -37.4 dB
[Auto Level] Current noise level -32.2 dB, estimated noise -31.5 dB
```

On this note, what is a good reception range if things are set up properly. One temperature sensor is ~20-25m line of sight away and the SDR doesn't receive a single signal. I'd think this is not very far. Another indication my setup is not correct?

> You get temperature stabilization and a metal case. Nooelec also has some compact ones (NESDR Nano 3). Look for R820T (R860 is a newer name), the E4000 ones might be more complicated.

Thanks for this pointer!! The Nano 3 seems to be made to be plugged in directly. I love it. I'll give this a shot. Together with [this ](https://www.amazon.com/gp/product/B07CBR485W) 433 MHz 10dBi antenna. Just in case you've a better recommendation, pls let me know.

Thank you so much for helping me to debug this!!

**Reply by zuckschwerdt** [reply]
I'm not used to that color scheme. Maybe capture a short sample file (`-w myfile_433.92M_250k.cu8 -T 30s`) and drop on https://triq.org/pdv/ or also upload here as zip.

20m LOS should work. With clear LOS 50m are no problem. But it depends on sender strength and walls in between.

I wouldn't plug any SDR directly into a Raspi, too much unshielded noise.

**Comment by gdt** [comment]
I also suggest higher-quality dongles.  In addition to temperature stability and metal case, they are often better calibrated to frequency in the first place, and can receive weaker signals.   To me, Nooelec metal-case TCXO units and Blog v3 are the only way to go, although I bet there are other good ones I'm not aware of.

**Reply by eigenphase** [reply]
Thanks for another “thumbs up” at this.

I’ve also heard the included antennas could be really bad. My SDR even has F (?) connector.

I now ordered the nano 3 sdr and a specific 433MHz antenna with SMA connector, 3dBi, narrow bandwidth and <1.5 SWR. Hope I’ll get substantial improvements
